knox/auth.py

```diff
@@ -74,7 +74,13 @@ def authenticate_credentials(self, token):
     def renew_token(self, auth_token) -> None:
         current_expiry = auth_token.expiry
         new_expiry = timezone.now() + knox_settings.TOKEN_TTL
+
+        # Do not auto-renew tokens past AUTO_REFRESH_MAX_TTL.
+        if knox_settings.AUTO_REFRESH_MAX_TTL is not None:
+            new_expiry = min(new_expiry, auth_token.created + knox_settings.AUTO_REFRESH_MAX_TTL)
```

Comment:
I guess I don't understand the library enough to get what needs to be different here. This seems to me like an elegant solution to not extending the maximum time since creation date. Is there a different exception or http code or something you want to return to the user instead?

Reply:
I would like the error message to clearly state that the token couldn’t be auto refreshed due to the max ttl being reached yes.

While this code would work, it would be quite hard to understand what happened.

Remember the Zen of Python: explicit is better than implicit 😉

Reply:
Okay, but the current design deletes token objects once they expire. The error message does not currently distinguish between tokens that never existed and ones that once existed then expired. According to the docstring of `authenticate_credentials`: "Tokens that have expired will be deleted and skipped", using `_cleanup_token`.

So I think it's on you to provide a mechanism to explicitly distinguish these things.

Reply:
This is not really true.

Nothing prevents us from raising something like ˋAutoRefreshMaxTTLExpired` here and handling the exception in `authenticate_credentialsˋ in order to return a specific error message  with ˋ exceptions.AuthenticationFailed`.

Marking the token has expired and letting it continue its life until it reaches another step of the flow feels wrong to me, really.

Reply:
Reading the code again I realize we actually don’t want an error here as the authentication passed … Then I guess adding a log message stating that the max ttl was reached would be better already.

And maybe for the longer game we should review the way this is designed: decide if a token should be refreshed at the time we identify it is actually expired - this would give us more room for raising appropriate errors but this is definitely outside of the scope of this PR

Reply:
Totally fine adding a log message, but there is currently no logging whatsoever in the project, so I would wait for you or one of the other main developers to add that before trying to do it myself. My client is already using our fork of this in its current state, so I'm going to stop working on this PR for now. Please feel free to do anything you need to to merge this into the main codebase.

Reply:
AFAIK logging configuration is handled by Django.

To add a log line here you can just follow the usual way:

`logger = logging.getLogger(__name__)`

and then use the logger accordingly.

Reply:
Added a log message when the expiry is capped by AUTO_REFRESH_MAX_TTL.